### PR TITLE
[8.x] Add parent model key when upserting has many relations

### LINKED ISSUE #40049: The method upsert() does not add the parent id when inserting on the relationship
- Laravel Version: 8.75.0
- PHP Version: 8.0.13
- Database Driver & Version: PostgreSQL 13.4

### Description:

If you want to use the ```upsert()``` method to insert or update relationships, then it will not add the id of the parent model.
For example, a similar ```createMany()``` method adds the id of the parent model.

### Steps To Reproduce:

1. Create a Post model with one-to-many User relationships. Migration can look like this:
```php
Schema::create('posts', function (Blueprint $table) {
            $table->id();
            $table->foreignIdFor(User::class)
                ->index()
                ->constrained()
                ->onDelete('cascade');
});
```
2. Define relationships in models:

```php
class User extends Model
{
    public function posts(): HasMany
    {
        return $this->hasMany(Post::class);
    }
}

class Post extends Model
{
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }
}
``` 

3. Add multiple posts via relationships:
```php
$user = User::find($id);
$postsRequest = $request->all();

$user->posts()
     ->upsert($postsRequest, ['id']);
```

4. In response, you will receive an error:

```sql
SQLSTATE[23502]: Not null violation: 7 ERROR: null value in column "user_id" of relation "posts" violates not-null 
constraint DETAIL: Failing row contains...
```
And if you print out the INPUT data, you will see that there really is no "user_id" field.

### PR DESCRIPTION
**Pull request by mateusjunges**
This is an attempt to fix #40049. However, i don't exactly know how to test this. @paras-malhotra as you are the one who introduced `upsert`, i'd appreciate your help.

### CODE REVIEW

**Comment by mateusjunges** [comment]
Can someone restart the jobs please?
image

File: tests/Database/DatabaseEloquentHasManyTest.php [hunk 1]
```diff
@@ -184,6 +184,17 @@ public function testUpdateOrCreateMethodCreatesNewModelWithForeignKeySet()
         $this->assertInstanceOf(Model::class, $relation->updateOrCreate(['foo'], ['bar']));
     }
 
+    public function testUpsertMethodCorrectlySetsParentModeley()
```
Comment (Neol3108) [inline tests/Database/DatabaseEloquentHasManyTest.php:187]:
```suggestion
    public function testUpsertMethodCorrectlySetsParentModelKey()
```

**Comment by taylorotwell** [comment]
Marking as draft until reviewed by @paras-malhotra  and has tests.

**Comment by taylorotwell** [comment]
I'm honestly not sure support for relationship upserting has even been considered yet.

File: src/Illuminate/Database/Eloquent/Relations/HasMany.php [hunk 2]
```diff
@@ -46,4 +46,22 @@ public function match(array $models, Collection $results, $relation)
     {
         return $this->matchMany($models, $results, $relation);
     }
+
+    /**
+     *  Insert new records or update the existing ones.
+     *
+     * @param  array  $values
+     * @param $uniqueBy
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/HasMany.php:54]: invalid syntax

File: src/Illuminate/Database/Eloquent/Relations/HasMany.php [hunk 3]
```diff
@@ -46,4 +46,22 @@ public function match(array $models, Collection $results, $relation)
...
+     * @param $update
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/HasMany.php:55]: same

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Database/Eloquent/Relations/HasMany.php (hunk 3 above)
Comment (mateusjunges) [reply src/Illuminate/Database/Eloquent/Relations/HasMany.php]: @GrahamCampbell fixed, thanks!

**Comment by driesvints** [comment]
Is this PR still being worked on? Otherwise it's best to close this and send it in at a later time.

**Comment by mateusjunges** [comment]
@driesvints i think taylor wants a review from @paras-malhotra. For me you guys can already review it.

**Comment by taylorotwell** [comment]
I basically know for certain `upsert` was never intended to be used via relationships. Only for upserting directly on an Eloquent model. 

If you are wanting to add it here, what about the other relationship types? Would those need support?

**Comment by taylorotwell** [comment]
Closing pending inactivity.

**Comment by kasus** [comment]
Here is how I figured this:
```php
HasMany::macro('upsertRelated', function (array $values, $uniqueBy, $update = null) {
      data_set($values, '*.' . $this->getForeignKeyName(), $this->getParentKey());
      return $this->related->upsert($values, $uniqueBy, $update);
});
```

not proud of my english but I just wanted to prevent overriding of the QueryBuilder `upsert` function.

Also using the similar for BelongsToMany relation:
```php
BelongsToMany::macro('upsertPivot', function (array $values, $uniqueBy, $update = null) {
      data_set($values, '*.' . $this->getForeignPivotKeyName(), $this->getParent()->getKey());
      return $this->newPivotQuery()->upsert($values, $uniqueBy, $update);
});
```
I needed it mostly because `sync()` function on belongsToMany relation is not very effective with lots of pivot records and it also can't do increments and stuff like this:
```php
$user->recipes()->upsertPivot($recipes, ['recipe_id', 'user_id'], ['servings' => DB::raw('servings + values(`servings`)')]);
```

I can make a PR to core if it makes sense for framework.

**Comment by khsuzan** [comment]
> Here is how I figured this:
> 
> ```
> HasMany::macro('upsertRelated', function (array $values, $uniqueBy, $update = null) {
>       data_set($values, '*.' . $this->getForeignKeyName(), $this->getParentKey());
>       return $this->related->upsert($values, $uniqueBy, $update);
> });
> ```
> 
> not proud of my english but I just wanted to prevent overriding of the QueryBuilder `upsert` function.
> 
> Also using the similar for BelongsToMany relation:
> 
> ```
> BelongsToMany::macro('upsertPivot', function (array $values, $uniqueBy, $update = null) {
>       data_set($values, '*.' . $this->getForeignPivotKeyName(), $this->getParent()->getKey());
>       return $this->newPivotQuery()->upsert($values, $uniqueBy, $update);
> });
> ```
> 
> I needed it mostly because `sync()` function on belongsToMany relation is not very effective with lots of pivot records and it also can't do increments and stuff like this:
> 
> ```
> $user->recipes()->upsertPivot($recipes, ['recipe_id', 'user_id'], ['servings' => DB::raw('servings + values(`servings`)')]);
> ```
> 
> I can make a PR to core if it makes sense for framework.

You made my day. Thanks

**Pull request closed without merging**